Math 215 - HW3 Template - Due Friday, October 11, 2024 by 11:59 pm
45 points total

1.  Gambler's Ruin (15 points)

2.  Integration (10 points)

3.  DNA strings with menu loop (20 points)

Problem 1 - Gambler's Ruin  (15 points)
1.	Gambler’s Ruin.  

A.	Suppose a gambler makes a series of fair \$1 bets, starting with \$50, and continues to play until they either go broke or have \$250. What are the chances that they will go home with \$250, and how many bets might they expect to make before winning or losing?  Write a program that reads three values from the user: the initial stake (in my example \$50), the goal amount (in my example \$250), and the number of times we want to simulate the game.  Your program then simulates the game - betting until the player either reaches their set goal or loses all their money.  The simulation of one game (flip the coin until the player reaches the goal or goes broke) should be in a function named game(stake, goal)   This function should return two values, the number of bets placed during the game and whether or not the game ended in a win for the player.  After implementing game, do a Monte Carlo simulation to estimate the probability that the player will win, and the average number of bets they make before winning or losing.  Your MonteCarlo simulation should make calls to your function
game(stake, goal).
Part of the point of this exercise is to organize your code using functions.  You must use the function headers I have provided and implement the functions to make them do what the comment say they do.   You may NOT modify the main function I have provided - you must write the functions game and simulate so that they work correctly with the main function.

B.	Experiment with running your code see what patterns you can find.  How likely is it that the gambler will double their stake?  Triple?  Go from a stake of \$50 to \$250?  Does it matter how big the initial stake and goal are?  For example, how to the results compare when stake = \$10 and goal = \$20, versus when  stake = \$100 and goal = \$200?  What patterns can you see in the number of flips needed?  For this part of the assignment, write your answers in a text cell.

Grading:

game:  5 points

simulate: 5 points

findings - part B:   5 points

In [11]:
import random

def game(stake, goal):
  # Input parameters:  stake, the player's initial amount of money in dollars
  #                    goal, the dollar amount that the player wants to reach - game ends if goal is reached or surpassed
  #                    win_prob a float number between 0 and 1 given the probability of getting heads in the coin flip
  # Return values:  (n, x) where n is the number of times the coin was flipped and
  #                    x = 1 if the player reaches or exceeds their goal, x = 0 if the game ends with the player losing all their money and not reaching their goal
  # The game:  Repeatedly flip the coin.  If it lands on heads, the player recieves $1,  If it lands on tails, the plays loses $1
  # Play continues until the play runs out of money, or reaches or exceeds their goal.

  num_of_bets = 0

  if stake == 0:
    return num_of_bets, 0

  while stake > 0 and stake < goal:
    choice = random.choice(["Heads", "Tails"])
    if choice == "Heads":
      stake+=1
    else:
      stake-=1
    num_of_bets+=1

  if stake == 0:
    return num_of_bets, 0
  elif stake >= goal:
    return num_of_bets, 1 # returning number of bets made and returning 1 if player wins
                          # or 0 if he goes broke


def simulate(num_trials, stake, goal):
  # Input parameters:  num_trials, a positive int value for the number of times to repeat game
  #                    stake, the player's initial amount of money in dollars
  #                    goal, the dollar amount that the player wants to reach - game ends if goal is reached or surpassed
  #                    win_prob a float number between 0 and 1 given the probability of getting heads in the coin flip.  The default value is 0.5 for a fair coin
  # The simulation:  Repeat gameA num_trials many times and determine how likely it is that the player will win
  # Return values: (pw, enf)  where pw is the estimated probability that the player will win  = num_wins/num_trials
  #                    and enf is the estimated number of flips untils the game ends

  num_wins = 0
  num_of_bets = 0
  for i in range(num_trials):
    total_bets, win_or_loss = game(stake, goal)
    num_of_bets = num_of_bets + total_bets
    if win_or_loss == 1:
      num_wins = num_wins + 1

  pw = num_wins / num_trials # estimated prob that gambler wins
  enf = num_of_bets / num_trials # estimated number of flips until game ends whether he wins/loses
  #no specified requ. to round avg # of flips

  return pw, enf



def main():
  print("Welcome to the Gambler's Ruin Program")
  s = float(input("What is the player's original stake: "))
  g = float(input("What is the player's goal: "))
  n = int(input('How many times should we run the experiment? '))

  prob_w, num_games = simulate(n, s, g)
  print(f'Using {n:,} trials,  stake = {s}, and goal = {g} ')
  print(f'the probability of winning is {prob_w} and the average number of flips to end the game is {num_games}')

main()

Welcome to the Gambler's Ruin Program
What is the player's original stake: 100
What is the player's goal: 200
How many times should we run the experiment? 10000
Using 10,000 trials,  stake = 100.0, and goal = 200.0 
the probability of winning is 0.4932 and the average number of flips to end the game is 9998.0508


Your answers for problem 1B go here.

Analysis for 1B:

It seems that the probability of doubling the gambler's stake tapers off at around 0.49. The probability of tripling the gambler's stake tapers off at around 0.33. As one would expect, the size of the initial stake and hopeful goal obviously matter when speculating the probability of winning on each of these games. As the stake and goal get farther apart, the probability of winning goes down. One experiment I ran was with a stake of $10 and a goal of $20, resulting in a winning prob. of 0.5 and an average of 101 flips. On the other hand, a stake of $100 and a goal of $200 resulted in a prob of  0.49 but an average flip number of 9,998. Another experiment I ran was checking the probability of winning and average number of flips for a stake of $50 and a goal of $250. This resulted in a probability of 0.2008 and the average number of flips clocking in at around 10,100. The pattern would be that the percentage increase/change between a stake and goal across experiments will likely allow the winning prob. to be similar; the average number of flips to end the game, however, will largely depend on how large the stake and goal themselves are.These experiments were all done with a Monte Carlo simulation of 10,000 trials.

Problem 2 :  Numerical Integration - Trapezoid Rule.  (10 points)

In class on September 26 we did an example with numerical integration.  We implemented a left-hand Riemann sum to approximate the area under a curve.  For this assignment you should modify that code to implement the trapezoid rule for approximation an integral.  You can read about the trapezoid rule here.  (We will use a uniform grid)
https://en.wikipedia.org/wiki/Trapezoidal_rule
I have pasted the code from Sept 26 below.  You will add another function to implement the trapezoid rule, and then call the function and compare results with the left-hand sums.
Your only work in this problem is to implement the trapezoid_integral function.  Do not change the other code.


In [12]:
#Problem 2
import numpy as np

def numerical_integral(a, b, f, n):
  # This is the Left-hand Riemann Sum - do not change this
  dx = (b-a)/n
  answer = 0
  x = a
  for k in range(n):
    answer += f(x)*dx
    x += dx
  return answer

def trapezoid_integral(a, b, f, n):
  trapezoid_const = 0.5 # avoids magic vals
  prelim_trapezoid_area = trapezoid_const * (f(a) + f(b)) # taken from trap formula wiki
  # deals with fxn values for start and end bounds, allows for loop to start at 1

  bound_diff = (b - a) / n # width of interval

  for j in range(1, n): # sums result (height) of func at each trapezoid
    x = a + j * bound_diff # calculates x val relative to beginning 'a' bound and width
    prelim_trapezoid_area += f(x)

  return prelim_trapezoid_area * bound_diff # multiply summed heights from loop by (consistent) width

def f1(x):
  return x*x

def f2(x):
  return np.exp(-x*x)

def main():
  num_rect = 1000
  print(f'The integral of x^2 from x = {0} to x = {3} using {num_rect} rectangles is {numerical_integral(0, 3, f1, num_rect):.4f}')
  print(f'The integral of x^2 from x = {0} to x = {3} using {num_rect} trapezoids is {trapezoid_integral(0, 3, f1, num_rect):.4f}')
  print()
  print(f'The integral of e^(-x^2) from x = {0} to x = {3} using {num_rect} rectangles is {numerical_integral(0, 3, f2, num_rect):.4f}')
  print(f'The integral of e^(-x^2) from x = {0} to x = {3} using {num_rect} trapezoids is {trapezoid_integral(0, 3, f2, num_rect):.4f}')
  print()
  print(f'The integral of sin(x) from x = {0} to x = {np.pi:.4f} using {num_rect} rectangles is {numerical_integral(0, np.pi, np.sin, num_rect):.4f}')
  print(f'The integral of sin(x) from x = {0} to x = {np.pi:.4f} using {num_rect} trapezoids is {trapezoid_integral(0, np.pi, np.sin, num_rect):.4f}')

main()

The integral of x^2 from x = 0 to x = 3 using 1000 rectangles is 8.9865
The integral of x^2 from x = 0 to x = 3 using 1000 trapezoids is 9.0000

The integral of e^(-x^2) from x = 0 to x = 3 using 1000 rectangles is 0.8877
The integral of e^(-x^2) from x = 0 to x = 3 using 1000 trapezoids is 0.8862

The integral of sin(x) from x = 0 to x = 3.1416 using 1000 rectangles is 2.0000
The integral of sin(x) from x = 0 to x = 3.1416 using 1000 trapezoids is 2.0000


Problem 3.  (20 points)

Complete the menu-driven DNA exercises we worked on in class on Sept 26.  I included a few more options in the menu.  You must implement those options and adjust the calling code to enable the options.  Task 5 will be one that you design and implement.  Here you will need to describe it in the menu and write the code to implement the task.

Grading:  

Task 2 (4 points)

Task 3  (4 points)

Task 4  (4 points)

Task 5  (4 points)

Complete the code for perform_action (4 points)


In [17]:
def show_menu():
  # show a menu of options
  print('Select an option: ')
  print('1:  Generate a random dna string')
  print('2:  Compute the G-concentration of a given DNA string')
  print('3:  Find the location of the substring of length 10 in a given DNA string with the highest G-concentration')
  print('4:  Find the location of the longest consecutive sequence of T\'s in a given DNA string')
  print('5:  Produce the reversed (backwards) version of the dna string')
  print('q:  Quit this program')
  # show a menu of options

def get_choice():
  # return the user's choice of options
  answer = input('What do you choose?')
  return answer

def random_dna(n): #1
  return ''.join([random.choice('ACTG') for k in range(n)])

# Implement functions for choices 2, 3, 4, 5

def g_concentrate(dna_string): #2
  dna_string = dna_string.upper()
  num_of_g = dna_string.count('G')
  return num_of_g / len(dna_string)

def indexOf_g_concentrate_substring(dna_string): #3
  g_concentrate_values = []
  starting_indices_of_substrs = []
  for i in range(len(dna_string) - 9):
    dna_substr = dna_string[i:i+10]
    starting_indices_of_substrs.append(i)
    #length of index list and g_concentrate list should be the same as number of different
    #substrings created from dna_string should correspond to indices of beginning value that they were created with

    g_concentrate_substr = g_concentrate(dna_substr)
    g_concentrate_values.append(g_concentrate_substr)

  max_g_concentrate = g_concentrate_values[0]
  index_of_max = 0
  for j in range(len(g_concentrate_values)):
    if g_concentrate_values[j] > max_g_concentrate:
      max_g_concentrate = g_concentrate_values[j]
      index_of_max = j

  return starting_indices_of_substrs[index_of_max] # should correspond to index within original dna_string of beginning of substr with highest g-concentr.

  # logic for loop iteration length for substring loop below:
  # abcdefghijklmno - total is 15, stopping at index 5 (f) for last substr of 10
  # abcdefghijklmnopqrstuvwxy total is 25, stopping at index 15 (p)
  # abcdefghijklmnopqrstuvwxyz total is 26, stopping at index 16 (q)

def consecutive_t(dna_string): #4
  max_index = -1 #longest consecutive t's
  current_start = -1 #caching index of current t and possible consecutive ones after it
  max_length = 0 #caching longest set of t's
  current_length = 0 #variable for comparison to max length

  for i in range(len(dna_string)):
    if dna_string[i] == 'T':
      if current_length == 0: #instantiates t-counting
        current_start = i
      current_length += 1
      if current_length > max_length:
        max_length = current_length
        max_index = current_start
    else:
      current_length = 0  #reset current t-tracking variables if t not found
      current_start = -1
  return max_index

def reverseDNA(dna_string): #5
  reverse_result = ""

  for i in range(len(dna_string)):
    reverse_result = dna_string[i] + reverse_result #prepends each letter to result so that it iterates/inserts "backwards" by adding to front
  return reverse_result

def perform_action(a):
  # a is a string.  If a is "random" then return a random dna string of length 10
  if a == '1':
    answer = random_dna(10)
    print(answer)
    return answer
  elif a == '2':
    answer = g_concentrate(random_dna(45))
    print(answer)
    return answer
  elif a == '3':
    answer = indexOf_g_concentrate_substring(random_dna(45))
    print(answer)
    return answer
  elif a == '4':
    answer = consecutive_t(random_dna(45))
    print(answer)
    return answer
  elif a == '5':
    answer = reverseDNA(random_dna(45))
    print(answer)
    return answer
  else:
    print('Unrecognized action code')

def main():
  print('Welcome')
  show_menu()
  user_choice = get_choice()
  while user_choice[0] not in ['q', 'Q']:
    perform_action(user_choice)
    show_menu()
    user_choice = get_choice()
  print('Goodbye')

main()

# providing a dna string as input - you can call random_dna(n) or get user input, no print or input statements in tasks2-5


Welcome
Select an option: 
1:  Generate a random dna string
2:  Compute the G-concentration of a given DNA string
3:  Find the location of the substring of length 10 in a given DNA string with the highest G-concentration
4:  Find the location of the longest consecutive sequence of T's in a given DNA string
5:  Produce the reversed version of the dna string
q:  Quit this program
What do you choose?4
2
Select an option: 
1:  Generate a random dna string
2:  Compute the G-concentration of a given DNA string
3:  Find the location of the substring of length 10 in a given DNA string with the highest G-concentration
4:  Find the location of the longest consecutive sequence of T's in a given DNA string
5:  Produce the reversed version of the dna string
q:  Quit this program
What do you choose?4
3
Select an option: 
1:  Generate a random dna string
2:  Compute the G-concentration of a given DNA string
3:  Find the location of the substring of length 10 in a given DNA string with the highest G-c